### PR TITLE
Type safe identifiers [CR]

### PR DESCRIPTION
**Pull request by BevapDin**
~~(WIP only because I haven't changed the code in cata_tiles to compile with the changed interface.)~~ Compiles fine with tiles.

As pointed out in #2047, various identifiers are implemented as `std::string`. There are some typedefs at work that make it more readable, but the compiler can not enforce the proper usage of them.

This PR implements a very thin wrapper `string_id` and `int_id` that wraps the `std::string` / `int` into a class, thereby enforcing the type checking.
##### More type safety

A monster id can not be compared to an item id, you can not call a function with an item id as parameter when it requires a monster id, functions can be overloaded based on the id type (e.g. `add(trap_id)` and `add(item_id)`).

In this way it also enforces better documented code. One can no longer use `std::string` or `int` as id, but must use the proper type.

It also allows static checks, one can search for all X id literals in the code (they look like `X_id( "some text" )`), which are clearly separated from other string literals (by the `X_id` prefix).

The int id can be constructed from the string id, without the user knowing how it's done (e.g. which map is uses). That way only a single function (that takes the int id) is needed, the other version (that would accept the id as string) is not required at all:

``` C++
void add_trap( trap_id t );
// no add_trap( trap_str_id t );

add_trap( trap_str_id( "tr_bubble" ) ); // automatically converted to trap_id
```
##### Less dependencies in header

The id-types can be defined without including the header of the identified object: including the string_id / int_id header and forward declaring the identified type (e.g. `struct trap`) is enough to typedef (and use) the id type.

``` C++
#include "int_id.h"
#include "string_id.h"
struct trap; // is enough, including of trap.h not required
using trap_id = int_id<trap>;
using trap_str_id = string_id<trap>;

trap_str_id tr_xx( "tr_xx" ); // works without trap.h
trap_id tr_XX( trap_str_id ); // string-to-int works without trap.h
```

"trap.h" is only needed when the actual trap object itself is required.

The typedefs of the various id types could all go into a single global header (which would have no further includes beside the string_id and int_id headers).
##### Ids as the main interface to the objects they identify

Currently all ids are handled separate from the objects the refer to, there are global functions that translate the id to the object (e.g. `terfind`). With this change, everything goes through the two id classes. This creates some consistency. 
Currently the answers to the questions
- "How do I check whether a given id is valid?"
- "How do I get the object a given id refers to?"
  depend on what type of object one means (item type, monster type, furniture type, ...) - with the new id classes, it's always the same:
- "Use the `is_valid` function of your id."
- "Use the `obj` function of your id."

``` C++
// hard coded:
trap_str_id some_trap_id = trap_str_id( "tr_bubble" );
// or from a json:
trap_str_id some_trap_id = trap_str_id( jso.get_string( "trap" ) );
// or from some function:
trap_str_id some_trap_id = some_thing.get_its_trap();

// this does not require one to know where it comes from (e.g. `traplist` / `trapmap`)
// and it's the same for either id type (int or string).
const trap &bubble = some_trap_id.obj();
```

---

It can also be used for the dynamic initialization (see #11798), using the string_id class as center to define the load/unload/check/finalize functions:

``` C++
template<typename T>
class string_id {
    static void load( JsonObject &jso );
    static void unload();
};
// and the implementation goes in the object specific cpp file (e.g. trapdef.cpp):
template<>
void string_id<trap>::load( JsonObject &jso )
{
}
template<>
void string_id<trap>::unload()
{
}
```

Furthermore: storing in json can be templated like this:

``` C++
template<typename T>
int_id<T> JsonObject::get_id( ... ) {
    // Load the ident as string, convert it to an integer id
    return string_id( get_string( ... ) ).id();
}

template<typename T>
void JsonObject::member( ..., int_id<T> id ) {
    // Convert the integer id to a string id, store the string
    member( ...,  id.id().str() );
}
```

Those function can be used for _all_ objects that use the id classes.

### CODE REVIEW

**Comment by Zireael07** [comment]
Great to see the code improving!

**Comment by kevingranade** [comment]
Good idea, just going to take a while to review.

File: src/editmap.cpp [hunk 1]
```diff
@@ -1118,18 +1117,18 @@ int editmap::edit_trp()
         for ( int t = tshift; t <= tshift + tmax; t++ ) {
             mvwprintz(w_picktrap, t + 1 - tshift, 1, c_white, "%s", padding.c_str());
             if ( t < num_trap_types ) {
+                auto &tr = trap_id( t ).obj();
```
Comment (Coolthulhu) [inline src/editmap.cpp]: This part doesn't seem to work correctly. It adds `tr_null`, but cuts off the last trap on the list, so `trap_id( 1 )` probably is the id of `tr_null`. Might cause problems in some rare cases (legacy saves?).

**Comment by Coolthulhu** [comment]
OK, tested it all and everything seems to work correctly.
Code itself looks readable (more so than master branch equivalents) and straightforward to use.

**Pull request merged**